fl4health/feature_alignment/tab_features_info_encoder.py

```diff
@@ -0,0 +1,130 @@
+import json
+from typing import Dict, List, Optional, Union
+
+import pandas as pd
+from cyclops.process.feature.feature import TabularFeatures
+from flwr.common.typing import Scalar
+from sklearn.feature_extraction.text import CountVectorizer
+
+from fl4health.feature_alignment.tabular_feature import MetaData, TabularFeature
+from fl4health.feature_alignment.tabular_type import TabularType
+
+
+class TabularFeaturesInfoEncoder:
+    """
+    This class encodes all the information required to perform feature
+    alignment on tabular datasets.
+
+    Parameters
+    ----------
+    tabular_features: List[TabularFeature]
+        List of all tabular features.
+    tabular_targets: List[TabularFeature]
+        List of all targets. (Note: targets are not included in tabular_features)
+    """
+
+    def __init__(self, tabular_features: List[TabularFeature], tabular_targets: List[TabularFeature]) -> None:
+        self.tabular_features = sorted(tabular_features, key=TabularFeature.get_feature_name)
+        self.tabular_targets = sorted(tabular_targets, key=TabularFeature.get_feature_name)
+
+    def get_tabular_features(self) -> List[TabularFeature]:
+        return self.tabular_features
+
+    def get_tabular_targets(self) -> List[TabularFeature]:
+        return self.tabular_targets
+
+    def get_feature_columns(self) -> List[str]:
+        return sorted([feature.get_feature_name() for feature in self.tabular_features])
+
+    def get_target_columns(self) -> List[str]:
+        return sorted([target.get_feature_name() for target in self.tabular_targets])
+
+    def features_by_type(self, tabular_type: TabularType) -> List[TabularFeature]:
+        return sorted(
+            [feature for feature in self.tabular_features if feature.get_feature_type() == tabular_type],
+            key=TabularFeature.get_feature_name,
+        )
+
+    def type_to_features(self) -> Dict[TabularType, List[TabularFeature]]:
+        return {tabular_type: self.features_by_type(tabular_type) for tabular_type in TabularType}
+
+    def get_categories_list(self) -> List[MetaData]:
+        return [cat_feature.get_metadata() for cat_feature in self.features_by_type(TabularType.ORDINAL)]
+
+    def get_target_dimension(self) -> int:
+        # Return the dimension of the target array *after* feature alignment is performed.
+        dimension = 0
+        for target in self.tabular_targets:
+            dimension += target.get_metadata_dimension()
+        return dimension
+
+    @staticmethod
+    def _construct_tab_feature(
```

Comment:
I think, but I'm not 100% sure (so let me know if I'm wrong), but I think the _ prefix in python is typically reserved for "private/protected" object methods (i.e. internal calls). However, I may be wrong and that this is also conventional for static calls.

Reply:
Yes you are right, it is typically reserved for internal calls. I used it here because I thought of `_construct_tab_feature` as a helper method that is only used internally.